new_beginnings

mahrous-amer opened the pull request:
dropping spotipy

mahrous-amer committed:
Django,Gunicorn,NGINX,Postgresql using Docker
Serve Static Files by Nginx from Django using Docker

mahrous-amer committed:
no longer dependent on spotipy we display user.playlist(s)

mahrous-amer committed:
Update README.md

mahrous-amer committed:
fixes and adding class: Playlist Track audio_features_object recommendations_seed_object recommendations_object

mahrous-amer committed:
update

mahrous-amer committed:
Big Sean but Logic -ally Alright

nael-binary commented on docker-compose.yml:
```diff
@@ -1,15 +1,28 @@
 services:
+  nginx:
+    image: nginx:latest
+    container_name: ng01
```
lets give it a more meaningful name, like full name. 
```suggestion
    container_name: nginx
```

so commands like this one: `docker exec -ti ng01 /bin/sh` can be `docker exec -ti nginx /bin/sh`

also we are not going to have such naming structure when we want to replicate in the future.
so can we please drop the numbering and more full names please :D

vnealv replied:
lets give it a more meaningful name, like full name. 
```suggestion
    container_name: nginx
```

so commands like this one: `docker exec -ti ng01 /bin/sh` can be `docker exec -ti nginx /bin/sh`

also we are not going to have such naming structure when we want to replicate in the future.
so can we please drop the numbering and more full names please :D

nael-binary commented on docker-compose.yml:
```diff
@@ -1,15 +1,28 @@
     build: .
     container_name: dg01
-    command: bash -c "python manage.py makemigrations && python manage.py migrate && gunicorn my_django.wsgi -b 0.0.0.0:8000"
+    command: bash -c "sleep 7; python manage.py makemigrations && python manage.py migrate && gunicorn my_django.wsgi -b 0.0.0.0:8000"
```
hmm, why adding `sleep 7`?

vnealv commented on docker-compose.yml:
```diff
@@ -1,15 +1,28 @@
     build: .
     container_name: dg01
-    command: bash -c "python manage.py makemigrations && python manage.py migrate && gunicorn my_django.wsgi -b 0.0.0.0:8000"
+    command: bash -c "sleep 7; python manage.py makemigrations && python manage.py migrate && gunicorn my_django.wsgi -b 0.0.0.0:8000"
```
hmm, why adding `sleep 7` here?

mahrous-amer replied:
Yes, I figured out we can set a 'waitfor' so that should do just fine.
```Bash
environment:
  - WAIT_HOSTS: postgres:5432
```

vnealv commented on src/grinder/views.py:
```diff
@@ -1,7 +1,183 @@
\ No newline at end of file
+print('''"Real knowledge is to know the extent of one's ignorance."'''+'\n                       '+"~Confucius~")
+
+#  Client Keys
+CLIENT_ID= os.environ.get("CLIENT_ID")
+CLIENT_SECRET= os.environ.get("CLIENT_SECRET")
+
+# Spotify URLS
+SPOTIFY_AUTH_URL = "https://accounts.spotify.com/authorize"
+SPOTIFY_TOKEN_URL = "https://accounts.spotify.com/api/token"
+SPOTIFY_API_BASE_URL = "https://api.spotify.com"
+API_VERSION = "v1"
+SPOTIFY_API_URL = "{}/{}".format(SPOTIFY_API_BASE_URL, API_VERSION)
+
+# Server-side Parameters
+CLIENT_SIDE_URL =  os.environ.get("client_side_url")
+PORT = os.environ.get("port")
+REDIRECT_URI = "{}:{}/callback/q".format(CLIENT_SIDE_URL, PORT)
+SCOPE = "playlist-modify-public playlist-modify-private user-top-read playlist-read-collaborative playlist-read-private"
+STATE = ""
+SHOW_DIALOG_bool = True
+SHOW_DIALOG_str = str(SHOW_DIALOG_bool).lower()
```
I believe all these should be moved to config file. Maybe `src/my_django/settings.py`?
Except for `SCOPE` I think we should keep it here, as we might need to dynamically change the scope we require.

mahrous-amer replied:
Basically, configurations that will be altered in this file otherwise in `src/my_django/settings.py`.

vnealv commented on src/grinder/views.py:
```diff
@@ -1,7 +1,183 @@
+    # Update User tokens and expire date
+    user.access_token = access_token
+    user.refresh_token = refresh_token
+    user.expires_in = expires_in
```
where do we make these updates takes place in Database?
will it do an UPDATE query with each time we assign new value? if yes can we make them all go in one query?

vnealv replied:
also `auth` function should finish here, and redirects the user to other function, something like `ViewUser` or `HomeUser`

vnealv commented on src/grinder/views.py:
```diff
@@ -1,7 +1,183 @@
+SHOW_DIALOG_bool = True
+SHOW_DIALOG_str = str(SHOW_DIALOG_bool).lower()
+
+auth_query_parameters = {
+    "response_type": "code",
+    "redirect_uri": REDIRECT_URI,
+    "scope": SCOPE,
+    # "state": STATE,
+    "show_dialog": SHOW_DIALOG_str,
+    "client_id": CLIENT_ID
+}
```
I think this should be a function by itself, 

```suggestion
def spotify_auth_url:
    auth_query_parameters = {
        "response_type": "code",
        "redirect_uri": REDIRECT_URI,
        "scope": SCOPE,
        # "state": STATE,
        "show_dialog": SHOW_DIALOG_str,
        "client_id": CLIENT_ID
    }
    url_args = "&".join(["{}={}".format(key, quote(val)) for key, val in auth_query_parameters.items()])
    auth_url = "{}/?{}".format(SPOTIFY_AUTH_URL, url_args)
    
    return auth_url
```

we can even have a separate helper class/file so we just import and use here in views

mahrous-amer replied:
That looks nice ill make it happen : )

vnealv commented on src/grinder/views.py:
```diff
@@ -1,7 +1,183 @@
+            raw_password = form.cleaned_data.get('password1')
+            user = authenticate(request, username=user_name, password=raw_password)
+            print("user = form.authenticate()", user)
+            login(request, user)
+            return redirect ('GetUser')
```
if you logged in the user, why are we redirecting him to `GetUser`?
```suggestion
            return redirect (spotify_auth_url)
```

vnealv replied:
if you logged in the user, why are we redirecting him to `GetUser`?
```suggestion
            login(request, user)
            return redirect (spotify_auth_url)
```

vnealv commented on src/grinder/views.py:
```diff
@@ -1,7 +1,183 @@
+    user.refresh_token = refresh_token
+    user.expires_in = expires_in
+
+    # Auth Step 6: Use the access token to access Spotify API
+    authorization_header = {"Authorization": "Bearer {}".format(access_token)}
+
+    # Get profile data
+    Current_users_profile_APIendpoint = "{}/me".format(SPOTIFY_API_URL)
+    User_profile_APIendpoint = "{}/users/{}".format(SPOTIFY_API_URL, user.username)
```
can we set these as objects in the class, that basically gets populated once we obtain these information. Seems like only `authorization_header` that needs an obtainable value also the username in `User_profile_APIendpoint`
so maybe move them to a function that if you passed a value it sets it and if you didn't it gives you whats in it

vnealv commented on src/grinder/views.py:
```diff
@@ -1,7 +1,183 @@
+    profile_data = json.loads(profile_response.text)
+
+    # top 100 tracks
+    top50_Apiendpoint ="{}/top/tracks".format(Current_users_profile_APIendpoint)
```
we need a nicer way to list and call different API calls.
something like a dictionary; key value representation.

```
SPOTIFY_API_CALL = {'user_profile': "{}/me".format(SPOTIFY_API_URL), 'top_tracks': "{}/top/tracks".format(Current_users_profile_APIendpoint)}
```

vnealv replied:
even nicer a function:

```
def spotify_api_call(call, *argv):

    calls = {'user_profile': "{}/me", 'top_tracks': "{}/top/tracks",  'audio_features': "{}/audio-features/?ids={}",  'playlist': "{}/playlists", 'auth': "{}/?{}"}
    return calls[call].format(SPOTIFY_API_URL, argv)
```

vnealv commented on src/grinder/models.py:
```diff
@@ -1,3 +1,91 @@
+        return self.username
+
+    def spRefresh (access_token, refresh_token, expires_in):
+        self.access_token = access_token
+        self.refresh_token = refresh_token
+        self.expires_in = expires_in
```
we really need to find a way to apply these updates to database in one go. And I think I found it. you will need to call `save` on object

```suggestion
        self.access_token = access_token
        self.refresh_token = refresh_token
        self.expires_in = expires_in
        self.save()
```

vnealv commented on src/grinder/views.py:
```diff
@@ -1,7 +1,183 @@
+    user = User.objects.get(username=request.user.username)
+
+    # Update User tokens and expire date
+    user.access_token = access_token
+    user.refresh_token = refresh_token
+    user.expires_in = expires_in
```
in here you should be calling `spRefresh`
```suggestion
    user.spRefresh(access_token, refresh_token, expires_in)
```

vnealv commented on src/grinder/views.py:
```diff
@@ -1,7 +1,183 @@
+    top50_Apiendpoint ="{}/top/tracks".format(Current_users_profile_APIendpoint)
+    top50_response = requests.get(top50_Apiendpoint, headers=authorization_header)
+    top50 = json.loads(top50_response.text)
+    ids = ''
+    for item in top50["items"]:
+        x = item["id"]
+        ids += x + ','
```
```suggestion
        ids = "," . join(v["id"] for v in top50["items"].values)
```

vnealv commented on src/grinder/views.py:
```diff
@@ -1,7 +1,183 @@
+    playlists_response = requests.get(playlist_api_endpoint, headers=authorization_header)
+    playlist_data = json.loads(playlists_response.text)
+
+    # create user output
+    output = '''<blockquote class="blockquote">'''
+    for item in playlist_data["items"]:
+        playlist_names = item["name"]
+        herfs = item["external_urls"]
+        playlist_hrefs = herfs["spotify"]
+        output +=  '''<kbd> <a href="'''+ playlist_hrefs +'''">'''+playlist_names+'''</a> </kbd> <br>'''
```
this loop should be in the template itself. In the view we should not parse html output. We just validate and prepare response data. Template is the one who is parsing and doing transformations to generate html content from this data.

vnealv commented on src/my_django/urls.py:
```diff
@@ -14,13 +14,13 @@
 urlpatterns = [
-    path('', include('grinder.urls')),
-    path('grinder', include('grinder.urls')),
+    path('', views.GetUser, name='GetUser'),
```
hmm, I don't get why we link path `''` to `GetUser`?

vnealv commented on src/templates/base.html:
```diff
@@ -1,32 +1,64 @@
-<html lang="en" dir="ltr">
-  <head>
-    <meta charset="utf-8">
+<html lang="en">
+<head>
+  {% load static %}
+  {{ form.media }}
+  <title>spotify-grinder</title>
+  <meta charset="UTF-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1">
+  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css">
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.js"></script>
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.16.0/umd/popper.min.js"></script>
+  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.0/js/bootstrap.min.js"></script>
+</script>
+</head>
+<body>
```
this static header section, should be in a separate file, as we will rarely change this, and we will need it in every template we create

vnealv commented on src/templates/base.html:
```diff
@@ -1,32 +1,64 @@
+<body>
 
-    <title>Spotify-Grinder</title>
-    {% load static %}
-    {% load staticfiles %}
```
whats in these two variables?

vnealv commented on src/templates/sign_in.html:
```diff
@@ -0,0 +1,33 @@
+{% extends 'base.html' %}
```
see we are adding big part of a template that may differ from one to another

mahrous-amer committed:
clean up and adding spotify wrapper 

A method to flag when User.Refresh_token needs to be refreshed is needed 
in grinder.User
